core/container/avl/avl.odin

```diff
@@ -0,0 +1,598 @@
+// present, the existing node is returned un-altered.
+insert :: proc(t: ^Tree($Value), value: Value) -> (ok: bool, n: ^Node(Value), err: runtime.Allocator_Error) {
+	cur: ^Node(Value)
+	cur_ptr := &t._root
+	for cur_ptr^ != nil {
+		cur = cur_ptr^
+		cmp := t._cmp_fn(value, cur.value)
+		switch {
+		case cmp < 0:
+			cur_ptr = &cur._left
+		case cmp > 0:
+			cur_ptr = &cur._right
+		case:
+			return false, cur, nil
+		}
+	}
+
+	n = new(Node(Value), t._node_allocator) or_return
+	n.value = value
+	n._parent = cur
+	cur_ptr^ = n
+	tree_rebalance_after_insert(t, n)
+	t._size += 1
+
+	return true, n, nil
+}
+
+// remove removes a node or value from the tree, and returns true iff the
+// removal was successful.  While the node's value will be left intact,
+// the node itself will be freed via the tree's node allocator.
+remove :: proc {
+	remove_value,
+	remove_node,
+}
+
+// remove_value removes a value from the tree, and returns true iff the
+// removal was successful.  While the node's value will be left intact,
+// the node itself will be freed via the tree's node allocator.
+remove_value :: proc(t: ^Tree($Value), value: Value, call_on_remove: bool = true) -> bool {
+	n := find(t, value)
+	if n == nil {
+		return false
+	}
+	return remove_node(t, n, call_on_remove)
+}
+
+// remove_node removes a node from the tree, and returns true iff the
+// removal was successful.  While the node's value will be left intact,
+// the node itself will be freed via the tree's node allocator.
+remove_node :: proc(t: ^Tree($Value), node: ^Node(Value), call_on_remove: bool = true) -> bool {
+	if node._parent == node || (node._parent == nil && t._root != node) {
+		return false
+	}
+	defer {
+		if call_on_remove && t.on_remove != nil {
+			t.on_remove(node.value, t.on_remove_user_data)
+		}
+		free(node, t._node_allocator)
+	}
+
+	parent: ^Node(Value)
+	left_deleted: bool
+
+	t._size -= 1
+	if node._left != nil && node._right != nil {
+		parent, left_deleted = tree_swap_with_successor(t, node)
+	} else {
+		child := node._left
+		if child == nil {
+			child = node._right
+		}
+		parent = node._parent
+		if parent != nil {
+			if node == parent._left {
+				parent._left = child
+				left_deleted = true
+			} else {
+				parent._right = child
+				left_deleted = false
+			}
+			if child != nil {
+				child._parent = parent
+			}
+		} else {
+			if child != nil {
+				child._parent = parent
+			}
+			t._root = child
+			node_reset(node)
+			return true
+		}
+	}
+
+	for {
+		if left_deleted {
+			parent = tree_handle_subtree_shrink(t, parent, +1, &left_deleted)
+		} else {
+			parent = tree_handle_subtree_shrink(t, parent, -1, &left_deleted)
+		}
+		if parent == nil {
+			break
+		}
+	}
+	node_reset(node)
+
+	return true
+}
+
+// iterator returns a tree iterator in the specified direction.
+iterator :: proc "contextless" (t: ^Tree($Value), direction: Direction) -> Iterator(Value) {
+	it: Iterator(Value)
+	it._tree = t
+	it._direction = direction
+
```

Comment:
I changed this to call iterator_first, but the bool and a special case needs to remain in iterator_next so that the very first call to iterator_next returns the first node instead of advancing.

The bool was renamed to reflect this, and iterator_first was made private to remove a edge case that causes non-intuitive behavior.